core/shared/src/main/scala-3/monocle/internal/focus/features/GeneratorLoop.scala

```diff
@@ -48,11 +48,38 @@ private[focus] trait GeneratorLoop {
       case a: FocusAction.KeywordWithDefault => generateWithDefault(a)
     }
 
+  import scala.util.control.NonFatal
+
+  private val _arity2 = Set(
+    TypeRepr.of[Lens[*, *]],
+    TypeRepr.of[Iso[*, *]],
+    TypeRepr.of[Prism[*, *]],
+    TypeRepr.of[Optional[*, *]],
+    TypeRepr.of[Traversal[*, *]],
+    TypeRepr.of[function.Each[*, *]],
+    TypeRepr.of[PLens[*, *, *, *]],
+    TypeRepr.of[PIso[*, *, *, *]],
+    TypeRepr.of[PPrism[*, *, *, *]],
+    TypeRepr.of[POptional[*, *, *, *]],
+    TypeRepr.of[PTraversal[*, *, *, *]]
+  )
+
+  /** Whether the expected `andThen` method has 2 type parameters. */
+  private def arity2(term: Term): Boolean =
+    term.tpe.widen match {
+      case AppliedType(tpe, _) => _arity2.exists(tpe <:< _)
+    }
+
   private def composeOptics(lens1: Term, lens2: Term): FocusResult[Term] =
-    lens2.tpe.widen match {
-      // Won't yet work for polymorphism where A != B, nor for non-polymorphic optics Getter or Fold.
-      case AppliedType(_, List(_, toType2)) =>
-        Right(Select.overloaded(lens1, "andThen", List(toType2, toType2), List(lens2)))
-      case _ => FocusError.ComposeMismatch(lens1.tpe.show, lens2.tpe.show).asResult
+    try
+      lens2.tpe.widen match {
```

Comment:
We should be able to do something a _little_ cleverer than this. There's a couple of newer APIs in dotty which means it's a little easier.